### PR TITLE
Adding spell-check command to SDK

### PR DESCRIPTION
**Pull request by roysagi**
Adding spell-check command to SDK.

Note: this also required adding "pyspellchecker" package to requirements.

The change in content_creator came from the precommit hooks

fixes: https://github.com/demisto/etc/issues/23706

### CODE REVIEW

**Comment by ronykoz** [comment]
@guykeller let me know when you finish reviewing this

File: demisto_sdk/dev_tools/spell_checker.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        self.spellchecker = SpellChecker()
+
+    def run_spell_check(self):
+        """Runs spell-check on the given file.
+
+        Returns:
+            int. 0 if no problematic words found, 1 otherwise.
+        """
+        unknown_words = set([])
+        self.spellchecker.word_frequency.load_words(KNOWN_WORDS)
```
Comment (guykeller) [inline demisto_sdk/dev_tools/spell_checker.py]: This line simply making these words known to the spell checker, right? 
If so let's add an explaining comment to it

File: demisto_sdk/dev_tools/spell_checker.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        """
+        for key, value in yml_info.items():
+            if key in DISPLAYABLE_LINES:
+                for word in value.split():
+                    if word.isalpha() and self.spellchecker.unknown([word]):
+                        unknown_words.add(word)
+
+            else:
+                if isinstance(value, dict):
+                    if key != SCRIPT_ARGS:
```
Comment (guykeller) [inline demisto_sdk/dev_tools/spell_checker.py]: Could you please explain this condition?

File: demisto_sdk/common/constants.py [hunk 3]
```diff
@@ -690,3 +690,22 @@ class PB_Status:
...
+
+KNOWN_WORDS = [
+    "Phishing",
+    "remediate",
+    "playbook",
+    "firewalls",
+    "IOCs",
+    "Emails",
+    "Google"
+]
```
[the remark this answers is not in the document]
Comment (guykeller) [reply demisto_sdk/common/constants.py]: Let's discuss known words @roysagi @ronykoz

**Comment by guykeller** [comment]
@roysagi Let's further review the logic and understand how this is working at the moment

File: demisto_sdk/common/constants.py [hunk 4]
```diff
@@ -690,3 +690,22 @@ class PB_Status:
...
 }
+
+DISPLAYABLE_LINES = [
+    "description",
+    "name",
+    "display",
+    "comment"
+]
+
+SCRIPT_ARGS = 'scriptarguments'
```
[the remark this answers is not in the document]
Comment (roysagi) [reply demisto_sdk/common/constants.py]: done

File: demisto_sdk/dev_tools/spell_checker.py (hunk 1 above)
Comment (roysagi) [reply demisto_sdk/dev_tools/spell_checker.py]: done

File: demisto_sdk/common/constants.py [hunk 5]
```diff
@@ -690,3 +690,22 @@ class PB_Status:
     'ip': {'IP.Address'},
     'url': {'URL.Data'}
 }
+
+DISPLAYABLE_LINES = [
```
Comment (ronykoz) [inline demisto_sdk/common/constants.py]: add a comment here or think of a better name please

File: demisto_sdk/common/constants.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ronykoz) [reply demisto_sdk/common/constants.py]: i think it should be hardcoded in the class itself perhaps

File: demisto_sdk/common/constants.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ronykoz) [reply demisto_sdk/common/constants.py]: it should be a file with words like Demisto, Dbotscore,... and we should have an option to provide our own file using a param

File: demisto_sdk/core.py [hunk 6]
```diff
@@ -158,3 +162,15 @@ def create_content_artifacts(self, artifact_path, preserve_bundles):
...
                         f'{json.dumps(cc.long_file_names, indent=4)}')
+
+    def spell_check(self, path: str) -> int:
+        """Runs SpellCheck for spell-check command.
+
+        Args:
+            path (str): The path to the checked file.
+
+        Returns:
+          int. 0 if no problematic words found, 1 otherwise.
```
Comment (ronykoz) [inline demisto_sdk/core.py]: use boolean

File: demisto_sdk/core.py [hunk 7]
```diff
@@ -72,6 +74,8 @@ def parse_args(self):
                 self.secrets(is_circle=args.circle, white_list_path=args.whitelist)
             elif args.command == 'create':
                 self.create_content_artifacts(args.artifacts_path, args.preserve_bundles)
+            elif args.command == 'spell-check':
+                self.spell_check(path=args.path)
```
Comment (ronykoz) [inline demisto_sdk/core.py]: we are not doing anything with it now, please add a TODO comment to fix this

File: demisto_sdk/dev_tools/spell_checker.py [hunk 8]
```diff
@@ -0,0 +1,95 @@
+import yaml
+
+from spellchecker import SpellChecker
+from demisto_sdk.common.tools import print_error, print_color, LOG_COLORS
+from demisto_sdk.common.constants import DISPLAYABLE_LINES, SCRIPT_ARGS, KNOWN_WORDS
+
+
+class SpellCheck:
+    """
```
Comment (ronykoz) [inline demisto_sdk/dev_tools/spell_checker.py]: add an explanation for the class itself

File: demisto_sdk/dev_tools/spell_checker.py [hunk 9]
```diff
@@ -0,0 +1,95 @@
+import yaml
+
+from spellchecker import SpellChecker
```
Comment (ronykoz) [inline demisto_sdk/dev_tools/spell_checker.py]: move this to line 2, keep the other ones separated by a line

File: demisto_sdk/dev_tools/spell_checker.py (hunk 1 above)
Comment (ronykoz) [reply demisto_sdk/dev_tools/spell_checker.py]: maybe this should be done in the init

File: demisto_sdk/dev_tools/spell_checker.py [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+            int. 0 if no problematic words found, 1 otherwise.
+        """
+        unknown_words = set([])
+        self.spellchecker.word_frequency.load_words(KNOWN_WORDS)
+
+        if self.path.endswith('.md'):
+            with open(self.path, 'r') as md_file:
+                md_data = md_file.readlines()
+
+            self.check_md_file(md_data=md_data, unknown_words=unknown_words)
```
[the remark this answers is not in the document]
Comment (ronykoz) [reply demisto_sdk/dev_tools/spell_checker.py]: we should have a space both before and after

File: demisto_sdk/dev_tools/spell_checker.py [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+
+    def __init__(self, path: str):
+        self.path = path
+        self.spellchecker = SpellChecker()
+
+    def run_spell_check(self):
+        """Runs spell-check on the given file.
+
+        Returns:
+            int. 0 if no problematic words found, 1 otherwise.
```
Comment (ronykoz) [inline demisto_sdk/dev_tools/spell_checker.py]: it's python, use boolean

File: demisto_sdk/dev_tools/spell_checker.py [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+
+    def check_md_file(self, md_data, unknown_words):
+        """Runs spell check on .md file. Adds unknown words to given unknown_words set.
+
+        Args:
+            md_data (list): A line list of the .md file contents
+            unknown_words (set): A set of found unknown words
+
+        Returns:
+            None.
```
Comment (ronykoz) [inline demisto_sdk/dev_tools/spell_checker.py]: remove

File: demisto_sdk/dev_tools/spell_checker.py [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+
+        Args:
+            yml_info (Loader): A line list of the .md file contents
+            unknown_words (set): A set of found unknown words
+
+        Returns:
+            None.
+        """
+        for key, value in yml_info.items():
+            if key in DISPLAYABLE_LINES:
```
[the remark this answers is not in the document]
Comment (ronykoz) [reply demisto_sdk/dev_tools/spell_checker.py]: yes, other fields could be field names etc, which are not relevant

File: demisto_sdk/dev_tools/spell_checker.py (hunk 2 above)
Comment (ronykoz) [reply demisto_sdk/dev_tools/spell_checker.py]: scriptarguments is something we should not explore, @roysagi lets add a comment about this

File: demisto_sdk/dev_tools/spell_checker.py [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+                for word in value.split():
+                    if word.isalpha() and self.spellchecker.unknown([word]):
+                        unknown_words.add(word)
+
+            else:
+                if isinstance(value, dict):
+                    if key != SCRIPT_ARGS:
+                        self.check_yaml(self.spellchecker, value, unknown_words)
```
Comment (ronykoz) [inline demisto_sdk/dev_tools/spell_checker.py]: add blank line after

File: demisto_sdk/dev_tools/spell_checker.py [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+                    for sub_list in value:
+                        if isinstance(sub_list, dict):
+                            self.check_yaml(self.spellchecker, sub_list, unknown_words)
+
+    @staticmethod
+    def add_sub_parser(subparsers):
+        from argparse import ArgumentDefaultsHelpFormatter
+        description = """Run spell check on a given yml/md file. """
+        parser = subparsers.add_parser('spell-check', help=description, formatter_class=ArgumentDefaultsHelpFormatter)
+        parser.add_argument("-p", "--path", help="Specify path of yml/md file", required=True)
```
Comment (ronykoz) [inline demisto_sdk/dev_tools/spell_checker.py]: add empty line at the end of the file

File: demisto_sdk/dev_tools/spell_checker.py [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+                    if key != SCRIPT_ARGS:
+                        self.check_yaml(self.spellchecker, value, unknown_words)
+                elif isinstance(value, list):
+                    for sub_list in value:
+                        if isinstance(sub_list, dict):
+                            self.check_yaml(self.spellchecker, sub_list, unknown_words)
+
+    @staticmethod
+    def add_sub_parser(subparsers):
+        from argparse import ArgumentDefaultsHelpFormatter
```
Comment (ronykoz) [inline demisto_sdk/dev_tools/spell_checker.py]: why is this here?

File: demisto_sdk/dev_tools/spell_checker.py (hunk 1 above)
Comment (roysagi) [reply demisto_sdk/dev_tools/spell_checker.py]: moved this to the attributes.

File: demisto_sdk/dev_tools/spell_checker.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (roysagi) [reply demisto_sdk/dev_tools/spell_checker.py]: done

File: demisto_sdk/dev_tools/spell_checker.py [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+
+            self.check_yaml(yml_info=yml_info, unknown_words=unknown_words)
+
+        else:
+            print_error("The file {} is not supported for spell-check command.\n"
+                        "Only .yml or .md files are supported.".format(self.path))
+            return 1
+
+        if unknown_words:
+            print_error(u"Found the problematic words in {}:\n{}".format(self.path, '\n'.join(unknown_words)))
```
[the remark this answers is not in the document]
Comment (roysagi) [reply demisto_sdk/dev_tools/spell_checker.py]: done

File: demisto_sdk/dev_tools/spell_checker.py [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+            return 1
+
+        print_color("No problematic words found in {}".format(self.path), LOG_COLORS.GREEN)
```
[the remark this answers is not in the document]
Comment (roysagi) [reply demisto_sdk/dev_tools/spell_checker.py]: done

File: demisto_sdk/dev_tools/spell_checker.py [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+        return 0
+
+    def check_md_file(self, md_data, unknown_words):
```
[the remark this answers is not in the document]
Comment (roysagi) [reply demisto_sdk/dev_tools/spell_checker.py]: changed

File: demisto_sdk/dev_tools/spell_checker.py (hunk 12 above)
Comment (roysagi) [reply demisto_sdk/dev_tools/spell_checker.py]: done

File: demisto_sdk/dev_tools/spell_checker.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (roysagi) [reply demisto_sdk/dev_tools/spell_checker.py]: made the name more descriptive and added further explanation.

File: demisto_sdk/dev_tools/spell_checker.py (hunk 2 above)
Comment (roysagi) [reply demisto_sdk/dev_tools/spell_checker.py]: its the field name for argument for commands in playbooks. this line is ok, I'll just add an explanation.

File: demisto_sdk/dev_tools/spell_checker.py (hunk 14 above)
Comment (roysagi) [reply demisto_sdk/dev_tools/spell_checker.py]: done

File: demisto_sdk/dev_tools/spell_checker.py (hunk 16 above)
Comment (roysagi) [reply demisto_sdk/dev_tools/spell_checker.py]: it is the way every add_sub_parser function is built in every class in the sdk.

File: demisto_sdk/dev_tools/spell_checker.py (hunk 15 above)
Comment (roysagi) [reply demisto_sdk/dev_tools/spell_checker.py]: done

File: demisto_sdk/dev_tools/spell_checker.py [hunk 20]
```diff
@@ -1,95 +1,129 @@
...
+    "description",
+    "display",
+    "name",
+    "comment"
+]
 
 
 class SpellCheck:
-    """
+    """Preform a spell check on the given .yml or .md file.
```
Comment (guykeller) [inline demisto_sdk/dev_tools/spell_checker.py]: Perform*

File: demisto_sdk/dev_tools/spell_checker.py (hunk 16 above)
Comment (ronykoz) [reply demisto_sdk/dev_tools/spell_checker.py]: @liorblob is there a good reason for this?

File: demisto_sdk/dev_tools/spell_checker.py [hunk 21]
```diff
@@ -1,95 +1,129 @@
...
     def run_spell_check(self):
         """Runs spell-check on the given file.
 
         Returns:
-            int. 0 if no problematic words found, 1 otherwise.
+            bool. True if no problematic words found, False otherwise.
         """
-        unknown_words = set([])
-        self.spellchecker.word_frequency.load_words(KNOWN_WORDS)
+        # adding known words file if given
```
Comment (guykeller) [inline demisto_sdk/dev_tools/spell_checker.py]: Let's add that known words means they don't count as misspelled words

File: demisto_sdk/dev_tools/spell_checker.py [hunk 22]
```diff
@@ -1,95 +1,129 @@
...
-            path (str): The path to the current file being checked
+            path (str): The path to the current file being checked.
+            known_words (str): The path to a file containing known words.
+            spellchecker (SpellChecker): The spell-checking object.
+            unknown_words (set): A set of unknown words found in the given file.
     """
 
-    def __init__(self, path: str):
+    def __init__(self, path: str, known_words: str):
         self.path = path
```
Comment (guykeller) [inline demisto_sdk/dev_tools/spell_checker.py]: Let's change this name to checked_file_path

File: demisto_sdk/dev_tools/spell_checker.py [hunk 23]
```diff
@@ -1,95 +1,129 @@
...
         """
-        unknown_words = set([])
-        self.spellchecker.word_frequency.load_words(KNOWN_WORDS)
+        # adding known words file if given
+        if self.known_words:
+            known_words_file = open(self.known_words, 'r')
+            self.spellchecker.word_frequency.load_text_file(known_words_file)
 
+        # adding the KNOWN_WORDS to the spellchecker recognized words.
+        self.spellchecker.word_frequency.load_words(KNOWN_WORDS)
```
Comment (guykeller) [inline demisto_sdk/dev_tools/spell_checker.py]: Let's move this logic to an add_known_words helper function

File: demisto_sdk/dev_tools/spell_checker.py [hunk 24]
```diff
@@ -1,95 +1,129 @@
...
         if self.path.endswith('.md'):
             with open(self.path, 'r') as md_file:
-                md_data = md_file.readlines()
+                md_file_lines = md_file.readlines()
```
Comment (guykeller) [inline demisto_sdk/dev_tools/spell_checker.py]: Let's move these lines into check_md_file

File: demisto_sdk/dev_tools/spell_checker.py [hunk 25]
```diff
@@ -1,95 +1,129 @@
...
+
+            self.check_md_file(md_file_lines=md_file_lines)
 
-            self.check_md_file(md_data=md_data, unknown_words=unknown_words)
         elif self.path.endswith('.yml'):
             with open(self.path, 'r') as yaml_file:
                 yml_info = yaml.safe_load(yaml_file)
```
Comment (guykeller) [inline demisto_sdk/dev_tools/spell_checker.py]: Let's move these lines into check_yaml

File: demisto_sdk/dev_tools/spell_checker.py [hunk 26]
```diff
@@ -1,95 +1,129 @@
...
-
-        Returns:
-            None.
         """
+        # yml file is parsed as a large dictionary. Each key in the large dictionary
+        # has a value which can be comprised of a string, a sub-dictionary or a list of dictionaries.
+        # The following code separates the spell check to these cases.
         for key, value in yml_info.items():
-            if key in DISPLAYABLE_LINES:
+            # case 1: a String.
```
Comment (guykeller) [inline demisto_sdk/dev_tools/spell_checker.py]: Let's change to case 1:string value of a user visible field or something like that

File: demisto_sdk/dev_tools/spell_checker.py [hunk 27]
```diff
@@ -1,95 +1,129 @@
...
+            if key in USER_VISIBLE_LINE_KEYS:
+
+                # separate the string to individual words
                 for word in value.split():
+
+                    # Drop commas, full-stops and brackets.
+                    if word.endswith(',') or word.endswith('.') or word.endswith(')'):
```
Comment (guykeller) [inline demisto_sdk/dev_tools/spell_checker.py]: What if it ends with ":" or "-"?

File: demisto_sdk/dev_tools/spell_checker.py (hunk 16 above)
Comment (liorblob) [reply demisto_sdk/dev_tools/spell_checker.py]: it's not in every class, it was just in the linter... no actual reason for it to be in every class.

File: demisto_sdk/dev_tools/spell_checker.py [hunk 28]
```diff
@@ -1,95 +1,129 @@
...
                     if word.isalpha() and self.spellchecker.unknown([word]):
-                        unknown_words.add(word)
+                        self.unknown_words.add(word)
 
             else:
+                # case 2: a sub-dictionary
                 if isinstance(value, dict):
-                    if key != SCRIPT_ARGS:
-                        self.check_yaml(self.spellchecker, value, unknown_words)
+                    # ignore command arguments in playbooks - no need to check these.
```
Comment (guykeller) [inline demisto_sdk/dev_tools/spell_checker.py]: Let's add reference to "scriptarguments" here

File: demisto_sdk/dev_tools/spell_checker.py (hunk 22 above)
Comment (roysagi) [reply demisto_sdk/dev_tools/spell_checker.py]: done

File: demisto_sdk/dev_tools/spell_checker.py (hunk 21 above)
Comment (roysagi) [reply demisto_sdk/dev_tools/spell_checker.py]: done

File: demisto_sdk/dev_tools/spell_checker.py (hunk 24 above)
Comment (roysagi) [reply demisto_sdk/dev_tools/spell_checker.py]: done

File: demisto_sdk/dev_tools/spell_checker.py (hunk 26 above)
Comment (roysagi) [reply demisto_sdk/dev_tools/spell_checker.py]: done

File: demisto_sdk/dev_tools/spell_checker.py (hunk 27 above)
Comment (roysagi) [reply demisto_sdk/dev_tools/spell_checker.py]: :  I agree... but why -?

File: demisto_sdk/dev_tools/spell_checker.py (hunk 28 above)
Comment (roysagi) [reply demisto_sdk/dev_tools/spell_checker.py]: it's kind of long - I dont think we want to do it
scriptarguments:
      message-id:
        complex:
          root: Gmail
          filters:
          - - operator: isNotEmpty
              left:
                value:
                  simple: Gmail.Attachments
                iscontext: true
          transformers:
          - operator: GetFirstObject
          - operator: getField
            args:
              field:
                value:
                  simple: ID

File: demisto_sdk/dev_tools/spell_checker.py (hunk 16 above)
Comment (roysagi) [reply demisto_sdk/dev_tools/spell_checker.py]: ok then... I'll take it out

File: demisto_sdk/dev_tools/spell_checker.py (hunk 28 above)
Comment (guykeller) [reply demisto_sdk/dev_tools/spell_checker.py]: I meant we should add a small reference to what "scriptarguments" is

File: demisto_sdk/dev_tools/spell_checker.py [hunk 29]
```diff
@@ -0,0 +1,126 @@
...
+    "display",
+    "name",
+    "comment"
+]
+
+
+class SpellCheck:
+    """Perform a spell check on the given .yml or .md file.
+        Attributes:
+            path (str): The path to the current file being checked.
```
Comment (guykeller) [inline demisto_sdk/dev_tools/spell_checker.py]:
```suggestion
            checked_file_path (str): The path to the current file being checked.
```

File: demisto_sdk/dev_tools/spell_checker.py [hunk 30]
```diff
@@ -0,0 +1,126 @@
...
+            self.spellchecker.word_frequency.load_text_file(known_words_file)
+
+        # adding the KNOWN_WORDS to the spellchecker recognized words.
+        self.spellchecker.word_frequency.load_words(KNOWN_WORDS)
+        if self.checked_file_path.endswith('.md'):
+            self.check_md_file()
+
+        elif self.checked_file_path.endswith('.yml'):
+            with open(self.checked_file_path, 'r') as yaml_file:
+                yml_info = yaml.safe_load(yaml_file)
```
Comment (guykeller) [inline demisto_sdk/dev_tools/spell_checker.py]: Let's add this code to the check_yaml method

File: demisto_sdk/dev_tools/spell_checker.py [hunk 31]
```diff
@@ -0,0 +1,126 @@
...
+
+            self.check_yaml(yml_info=yml_info)
+
+        else:
+            print_error("The file {} is not supported for spell-check command.\n"
+                        "Only .yml or .md files are supported.".format(self.checked_file_path))
+            return False
+
+        if self.unknown_words:
```
Comment (guykeller) [inline demisto_sdk/dev_tools/spell_checker.py]:
```suggestion
        if len(self.unknown_words) > 0:
```

File: demisto_sdk/dev_tools/spell_checker.py [hunk 32]
```diff
@@ -0,0 +1,126 @@
...
+                # case 2: a sub-dictionary
+                if isinstance(value, dict):
+                    # ignore command arguments in playbooks - no need to check these.
+                    if key != 'scriptarguments':
+                        self.check_yaml(value)
+
+                # case 3: a list of dictionaries
+                elif isinstance(value, list):
+                    for sub_dict in value:
+                        if isinstance(sub_dict, dict):
```
Comment (guykeller) [inline demisto_sdk/dev_tools/spell_checker.py]: Are we sure all of the sub_dicts will be a dict and none will be a list?

File: demisto_sdk/dev_tools/spell_checker.py (hunk 29 above)
Comment (roysagi) [reply demisto_sdk/dev_tools/spell_checker.py]: done

File: demisto_sdk/dev_tools/spell_checker.py (hunk 30 above)
Comment (roysagi) [reply demisto_sdk/dev_tools/spell_checker.py]: I don't think we want to do that. The check_yml function is recursive and works by getting a dictionary.
This is simply the initialisation of the first dictionary, putting it there will add an unnecessary step in my opinion.

File: demisto_sdk/dev_tools/spell_checker.py (hunk 31 above)
Comment (roysagi) [reply demisto_sdk/dev_tools/spell_checker.py]: done

File: demisto_sdk/dev_tools/spell_checker.py (hunk 28 above)
Comment (roysagi) [reply demisto_sdk/dev_tools/spell_checker.py]: done

File: demisto_sdk/dev_tools/spell_checker.py (hunk 32 above)
Comment (roysagi) [reply demisto_sdk/dev_tools/spell_checker.py]: I parsed both an integration and a playbook.
In an integration, I didn't find '[[' - so no nested lists as far as I can tell.
In a playbook the only '[[' I found were for operators and thats under "scriptarguments" which we ignore - like so:
'**scriptarguments**': {'format': {}, 'thread-id': {'complex': {'root': 'Gmail', 'filters': **[[**{'operator': 'isEqualString', 'left': {'value': {'simple': 'Gmail.Mailbox'}, 'iscontext': True}, 'right': {'value': {'simple': 'admin@demistodev.com'}}}]]...

**Comment by roysagi** [comment]
Added nltk package (https://www.nltk.org/).
Specifically I added the 'words' sub-package which is downloaded to the user's computer upon running spellcheck (so the first run is a little slower but not by too much).

There is a bigger dictionary ('brown') but it's about 8G so it seems a bit excessive - but we can add this to first installation and then we can use it.

File: demisto_sdk/dev_tools/spell_checker.py [hunk 33]
```diff
@@ -0,0 +1,148 @@
...
+    "name",
+    "comment"
+]
+
+
+class SpellCheck:
+    """Perform a spell check on the given .yml or .md file.
+        Attributes:
+            checked_file_path (str): The path to the current file being checked.
+            known_words (str): The path to a file containing known words.
```
Comment (ronykoz) [inline demisto_sdk/dev_tools/spell_checker.py]: known_words_file_path

File: demisto_sdk/dev_tools/spell_checker.py [hunk 34]
```diff
@@ -0,0 +1,148 @@
...
+            spellchecker (SpellChecker): The spell-checking object.
+            unknown_words (set): A set of unknown words found in the given file.
+    """
+
+    def __init__(self, checked_file_path: str, known_words: str):
+        self.checked_file_path = checked_file_path
+        self.spellchecker = SpellChecker()
+        self.unknown_words = set([])  # type:Set
+        self.known_words = known_words
```
Comment (ronykoz) [inline demisto_sdk/dev_tools/spell_checker.py]: same_here

File: demisto_sdk/dev_tools/spell_checker.py [hunk 35]
```diff
@@ -0,0 +1,148 @@
...
+        # adding the KNOWN_WORDS to the spellchecker recognized words.
+        self.spellchecker.word_frequency.load_words(KNOWN_WORDS)
+
+        # disabling SSL connection for nltk download.
+        try:
+            _create_unverified_https_context = ssl._create_unverified_context
+        except AttributeError:
+            pass
+        else:
+            ssl._create_default_https_context = _create_unverified_https_context
```
Comment (ronykoz) [inline demisto_sdk/dev_tools/spell_checker.py]: what is happening here

File: demisto_sdk/dev_tools/spell_checker.py [hunk 36]
```diff
@@ -0,0 +1,148 @@
...
+
+        # downloading words set from nltk.
+        nltk.download('words')
```
Comment (ronykoz) [inline demisto_sdk/dev_tools/spell_checker.py]: why do we need this if we have the corpus of words?

File: demisto_sdk/common/constants.py [hunk 37]
```diff
@@ -695,3 +695,13 @@ class PB_Status:
     'ip': {'IP.Address'},
     'url': {'URL.Data'}
 }
+
+KNOWN_WORDS = [
```
Comment (ronykoz) [inline demisto_sdk/common/constants.py]: we should have this as a file and read it, add words like Demisto, DBotScore please

File: demisto_sdk/dev_tools/spell_checker.py [hunk 38]
```diff
@@ -0,0 +1,148 @@
...
+            return False
+
+        print_color("\nNo misspelled words found in {}".format(self.checked_file_path), LOG_COLORS.GREEN)
+        return True
+
+    def add_known_words(self):
+        # adding known words file if given - these words will not count as misspelled
+        if self.known_words:
+            known_words_file = open(self.known_words, 'r')
+            self.spellchecker.word_frequency.load_text_file(known_words_file)
```
Comment (ronykoz) [inline demisto_sdk/dev_tools/spell_checker.py]: use with statement please
you are not closing the file and it will close it automatically when using this

File: demisto_sdk/common/constants.py (hunk 37 above)
Comment (roysagi) [reply demisto_sdk/common/constants.py]: moved out of constants to a different file

File: demisto_sdk/dev_tools/spell_checker.py (hunk 35 above)
Comment (roysagi) [reply demisto_sdk/dev_tools/spell_checker.py]: added a description for this move as well as the download

File: demisto_sdk/dev_tools/spell_checker.py (hunk 36 above)
Comment (roysagi) [reply demisto_sdk/dev_tools/spell_checker.py]: see above

File: demisto_sdk/dev_tools/spell_checker.py [hunk 39]
```diff
@@ -0,0 +1,150 @@
...
+                    for sub_dict in value:
+                        if isinstance(sub_dict, dict):
+                            self.check_yaml(sub_dict)
+
+    @staticmethod
+    def add_sub_parser(subparsers):
+        description = """Run spell check on a given yml/md file. """
+        parser = subparsers.add_parser('spell-check', help=description, formatter_class=ArgumentDefaultsHelpFormatter)
+        parser.add_argument("-p", "--path", help="Specify path of yml/md file", required=True)
+        parser.add_argument("--known_words", help="A file path to a txt file with known words")
```
Comment (ronykoz) [inline demisto_sdk/dev_tools/spell_checker.py]:
```suggestion
        parser.add_argument("--known_words", help="A file path to a txt file with known words to exclude in the check")
```

**Comment by roysagi** [comment]
Will add Unit tests before merging.

File: requirements.txt [hunk 40]
```diff
@@ -1,10 +1,13 @@
 bandit>=1.6.2
 bs4>=0.0.1
+bs4>=0.0.1
```
Comment (ronykoz) [inline requirements.txt]: you have duplicates

File: docs/spell_check_command.md [hunk 41]
```diff
@@ -0,0 +1,24 @@
+### Spell-check
```
Comment (ronykoz) [inline docs/spell_check_command.md]: @dantavori apply your style changes here too please

**Comment by ronykoz** [comment]
@yaakovi maybe we should add this one?

**Comment by Itay4** [comment]
@ronykoz is this still relevant ?
